File: src/main/java/com/leijendary/spring/boot/template/config/JpaConfiguration.java
package com.leijendary.spring.boot.template.config;

import static com.leijendary.spring.boot.template.util.RequestContext.now;
import static java.util.Optional.of;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.auditing.DateTimeProvider;
import org.springframework.data.jpa.repository.config.EnableJpaAuditing;

@Configuration
@EnableJpaAuditing(dateTimeProviderRef = "dateTimeProvider")
public class JpaConfiguration {

    @Bean
    public DateTimeProvider dateTimeProvider() {
        return () -> of(now());
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/config/JpaConfiguration.kt
package com.leijendary.spring.template.core.config

import com.leijendary.spring.template.core.util.RequestContext.now
import com.leijendary.spring.template.core.util.RequestContext.userId
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.auditing.DateTimeProvider
import org.springframework.data.domain.AuditorAware
import org.springframework.data.jpa.repository.config.EnableJpaAuditing
import org.springframework.transaction.annotation.EnableTransactionManagement
import java.util.Optional.of

@Configuration
@EnableJpaAuditing(auditorAwareRef = "auditorProvider", dateTimeProviderRef = "dateTimeProvider")
@EnableTransactionManagement
class JpaConfiguration {
    @Bean
    fun auditorProvider(): AuditorAware<String> {
        return AuditorAware { of(userId) }
    }

    @Bean
    fun dateTimeProvider(): DateTimeProvider {
        return DateTimeProvider { of(now) }
    }
}
